### PR TITLE
Feature/cache gl state

### PR DESCRIPTION
**Pull request by Mee-gu**
cache GL state

### CODE REVIEW

File: cocos/renderer/backend/opengl/GLStateCached.h [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        GLenum dstRGBBlendFactor = GL_ZERO;
+        GLenum dstALphaBlendFactor = GL_ZERO;
+
+        GLenum bufferTarget = GL_ARRAY_BUFFER;
+        GLuint bufferObject = GL_ZERO;
+        GLuint framebufferObject = GL_ZERO;
+
+        GLenum textureUnit = GL_TEXTURE0;
+        GLenum textureTarget = GL_TEXTURE_2D;
+        GLuint textureObject = GL_ZERO;
```
Comment (minggo) [inline cocos/renderer/backend/opengl/GLStateCached.h:85]: It is better to query the state at start. The driver may not obey the spec.

File: cocos/renderer/backend/opengl/GLStateCached.cpp [hunk 2]
```diff
@@ -0,0 +1,386 @@
+#include "GLStateCached.h"
+CC_BACKEND_BEGIN
+
+#define GL_STATE_CACHED_ENABLED
+#ifdef GL_STATE_CACHED_ENABLED
+#define RETURN() return;
```
Comment (minggo) [inline cocos/renderer/backend/opengl/GLStateCached.cpp:6]: return; -> return

**Comment by crazyhappygame** [comment]
Is that really needed?
What is the speedup?

**Comment by Mee-gu** [comment]
> Is that really needed?
> What is the speedup?

@crazyhappygame  GLStateCached will record the last-set value for mostly graphics API state, it will eliminate redundant API calls(e.g. enabling a state when it's already enabled). It's unnecessary to set these state repeatedly, although some calls are quick.

**Comment by crazyhappygame** [comment]
@Mee-gu thank you for explanation.
I got an idea.
I am curious how much this code improve performance.

**Comment by minggo** [comment]
I am not sure how much performance is improved with this PR too. It seems it depends on the implementation of GL driver implementation.
 
@Mee-gu i think we should do test on some android devices to make confident. The test doesn't have to use the engine, just use pure GL codes.

**Pull request closed without merging**